pkg/controller/utils/elasticsearch.go

```diff
@@ -373,3 +376,75 @@ func getTotalEsDisk(ls *operatorv1.LogStorage) int64 {
 	}
 	return totalEsStorage
 }
+
+// EnsureCertificateKeySecret ensures that the certificate in the provided
+// secrets has the expected DNS names. If no key secret is provided, a new key
+// secret is created and returned. If the key secret provided does have the
+// right DNS name, then that given key secret is returned.
+// Otherwise a new key secret is created and returned.
+func EnsureCertificateKeySecret(ctx context.Context, keySecretName string, keySecret *corev1.Secret, pubSecret *corev1.Secret, svcDNSNames ...string) (*corev1.Secret, error) {
+	var err error
+	secret := &corev1.Secret{}
+
+	// Create the key secret if it doesn't exist.
+	if keySecret == nil {
+		secret, err = render.CreateOperatorTLSSecret(nil,
+			keySecretName, "tls.key", "tls.crt",
+			render.DefaultCertificateDuration, nil, svcDNSNames...,
+		)
+		if err != nil {
+			return nil, err
+		}
+		return secret, nil
+	}
+
+	// Now check to see if the cert exists. If the cert's DNS
+	// names have changed then we need to recreate the key secret.
+	// If we don't have a pub secret yet then just return the key secret
+	if pubSecret == nil {
+		return secret, nil
+	}
+
+	// If we do have the pub secret then we need to check that its DNS names are expected
+	ok, err := secretHasExpectedDNSNames(pubSecret, svcDNSNames)
+
+	if err != nil {
+		return nil, err
+	}
+	// DNS names on the cert do not match expected values; create a new cert.
+	if !ok {
+		return render.CreateOperatorTLSSecret(nil,
+			keySecret.Name, "tls.key", "tls.crt",
+			render.DefaultCertificateDuration, nil, svcDNSNames...,
+		)
+	}
+
+	// Finally return just the key secret.
+	return secret, nil
+}
+
+func secretHasExpectedDNSNames(secret *corev1.Secret, expectedDNSNames []string) (bool, error) {
+	if secret.Data == nil {
+		return false, nil
+	}
+
+	certBytes := secret.Data["tls.crt"]
+	pemBlock, _ := pem.Decode(certBytes)
+	cert, err := x509.ParseCertificate(pemBlock.Bytes)
+	if err != nil {
+		return false, err
+	}
+
+	return reflect.DeepEqual(cert.DNSNames, expectedDNSNames), nil
+}
+
+func GetCertificateSecret(ctx context.Context, client client.Client, name string, ns string) (*corev1.Secret, error) {
```

Comment:
This doesn't really have anything to do with specifically getting a secret that contains a certificate, it's just shortcutting the code to get a secret in general.

Reply:
Moved to its own file